src/main/java/net/minecraftforge/fmlclient/StartupMessageRenderer.java

```diff
@@ -0,0 +1,279 @@
+            #version 150 core
+
+            in vec2 position;
+
+            uniform vec2 posMultiplier;
+            uniform vec2 posShift;
+            uniform float size;
+
+            void main() {
+                gl_Position = vec4((((position * size) + posShift) * posMultiplier) - 1.0, 0.0, 1.0);
+                gl_Position.y *= -1.0;
+            }
+            """;
+        
+        private static final String fragmentShaderSource = """
+            #version 150 core
+            
+            uniform vec4 colorIn;
+
+            out vec4 color;
+
+            void main() {
+                color = colorIn;
+            }
+            """;
+        
+        private final int program;
+        private final int posMultiplierLocation;
+        private final int posShiftLocation;
+        private final int colorInLocation;
+        private final int sizeLocation;
+        
+        private final int VAO;
+        
+        private final int dataBuffer;
+        private int bufferLength = 27000;
+        private ByteBuffer charBuffer;
+        
+        private int quadCount = 0;
+        private final int elementBuffer;
+        
+        private float height;
+        
+        public StartupMessageRenderer() {
+            int previousVAO = glGetInteger(GL_VERTEX_ARRAY_BINDING);
+            int previousArrayBuffer = glGetInteger(GL_ARRAY_BUFFER_BINDING);
+            int previousElementBuffer = glGetInteger(GL_ELEMENT_ARRAY_BUFFER_BINDING);
+            
+            int vertexShader = glCreateShader(GL_VERTEX_SHADER);
+            int fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+            
+            glShaderSource(vertexShader, vertexShaderSource);
+            glShaderSource(fragmentShader, fragmentShaderSource);
+            
+            glCompileShader(vertexShader);
+            glCompileShader(fragmentShader);
+            
+            if (glGetShaderi(vertexShader, GL_COMPILE_STATUS) != GL_TRUE || glGetShaderi(fragmentShader, GL_COMPILE_STATUS) != GL_TRUE) {
+                System.err.println("Shader compilation failed");
+                System.err.println("Vertex shader log\n" + glGetShaderInfoLog(vertexShader));
+                System.err.println("Fragment shader log\n" + glGetShaderInfoLog(fragmentShader));
+                throw new IllegalStateException("Shader compilation failed");
+            }
+            
+            program = glCreateProgram();
+            glAttachShader(program, vertexShader);
+            glAttachShader(program, fragmentShader);
+            glLinkProgram(program);
+            
+            if (glGetProgrami(program, GL_LINK_STATUS) != GL_TRUE) {
+                System.err.println("Program linking failed");
+                System.err.println("Program link log\n" + glGetProgramInfoLog(program));
+                throw new IllegalStateException("Program link failed");
+            }
+            
+            glDeleteShader(vertexShader);
```

Comment:
Shaders need to be detached (`glDetachShader`) before they can be deleted

Reply:
[https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDeleteShader.xhtml](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDeleteShader.xhtml)

> If a shader object to be deleted is attached to a program object, it will be flagged for deletion, but it will not be deleted until it is no longer attached to any program object, for any rendering context (i.e., it must be detached from wherever it was attached before it will be deleted). A value of 0 for shader will be silently ignored.

[https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDeleteProgram.xhtml](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDeleteProgram.xhtml)

> If a program object is in use as part of current rendering state, it will be flagged for deletion, but it will not be deleted until it is no longer part of current state for any rendering context. If a program object to be deleted has shader objects attached to it, those shader objects will be automatically detached but not deleted unless they have already been flagged for deletion by a previous call to glDeleteShader. A value of 0 for program will be silently ignored.

This is valid and defined to do.

Reply:
Comment was actually regarding why i dont detach the objects after linking, so OpenGL can immediately delete the objects instead of doing so when the program is later deleted. _I forgot you can do that._